server/src/main/java/org/elasticsearch/TransportVersion.java

```diff
@@ -229,20 +227,48 @@ public static TransportVersion max(TransportVersion version1, TransportVersion v
     }
 
     /**
-     * returns a sorted set of all transport version constants
+     * returns a sorted collection of declared transport version constants
      */
-    public static NavigableSet<TransportVersion> getAllVersions() {
-        return ALL_VERSIONS;
+    public static Collection<TransportVersion> getAllVersions() {
+        return VERSION_IDS.values();
     }
 
-    final int id;
+    public final int id;
     private final String uniqueId;
 
     TransportVersion(int id, String uniqueId) {
         this.id = id;
         this.uniqueId = Strings.requireNonEmpty(uniqueId, "Each TransportVersion needs a unique string id");
     }
 
+    /**
+     * Placeholder method for code compatibility with code calling {@code CURRENT.minimumCompatibilityVersion}.
+     */
+    @Deprecated
+    public TransportVersion minimumCompatibilityVersion() {
+        assert this.equals(CURRENT);
+        return MINIMUM_COMPATIBLE;
+    }
+
+    @Deprecated
+    public boolean isCompatible(TransportVersion version) {
+        return onOrAfter(version.calculateMinimumCompatVersion()) && version.onOrAfter(calculateMinimumCompatVersion());
+    }
+
+    private TransportVersion minimumCompatibleVersion;
+
+    /**
+     * Placeholder for code calling {@code minimumCompatibilityVersion} on arbitrary Version instances.
+     * Code calling this should be refactored to not do this.
+     */
+    @Deprecated
+    public TransportVersion calculateMinimumCompatVersion() {
```

Comment:
This is the 'complex' compatibility method, for callers actually requiring minimum compatibility for previous versions. This is largely handshake code and some tests. Callers will need to be refactored later on to not require this.